src/core/src/op/shape_of.cpp

```diff
@@ -170,77 +109,89 @@ bool evaluate_label(const Node* shape_of_node, TensorLabelVector& output_labels)
 }  // namespace
 }  // namespace shape_of
 
-bool op::v3::ShapeOf::evaluate(const HostTensorVector& output_values, const HostTensorVector& input_values) const {
-    OV_OP_SCOPE(v3_ShapeOf_evaluate);
-    OPENVINO_SUPPRESS_DEPRECATED_START
-    OPENVINO_ASSERT(validate_host_tensor_vector(input_values, 1));
-    OPENVINO_ASSERT(validate_host_tensor_vector(output_values, 1));
-    OPENVINO_SUPPRESS_DEPRECATED_END
-    return shape_of::evaluate_shape_of(output_values[0], input_values[0]);
+namespace v3 {
+ShapeOf::ShapeOf(const Output<Node>& arg, element::Type output_type) : ShapeOfBase({arg}), m_output_type(output_type) {
+    constructor_validate_and_infer_types();
+}
+
+void ShapeOf::validate_and_infer_types() {
+    OV_OP_SCOPE(v3_ShapeOf_validate_and_infer_types);
+    NODE_VALIDATION_CHECK(this,
+                          m_output_type == element::i64 || m_output_type == element::i32,
+                          "Output type must be i32 or i64");
+    set_input_is_relevant_to_value(0, false);
+    const auto input_partial_shape = get_input_partial_shape(0);
+    set_output_type(0, m_output_type, PartialShape{input_partial_shape.rank()});
+}
+
+bool ShapeOf::visit_attributes(AttributeVisitor& visitor) {
+    OV_OP_SCOPE(v3_ShapeOf_visit_attributes);
+    visitor.on_attribute("output_type", m_output_type);
+    return true;
+}
+
+std::shared_ptr<Node> ShapeOf::clone_with_new_inputs(const OutputVector& new_args) const {
+    OV_OP_SCOPE(v3_ShapeOf_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    return std::make_shared<ShapeOf>(new_args.at(0), m_output_type);
 }
 
-bool op::v3::ShapeOf::evaluate(ov::TensorVector& output_values, const ov::TensorVector& input_values) const {
+bool ShapeOf::evaluate(TensorVector& output_values, const TensorVector& input_values) const {
     OV_OP_SCOPE(v0_ShapeOf_evaluate);
-    OPENVINO_ASSERT(input_values.size() == 1);
-    OPENVINO_ASSERT(output_values.size() == 1);
+    OPENVINO_ASSERT(input_values.size() == 1 && output_values.size() == 1);
```

Comment:
The disadvantage of one assert is lack of information which size is wrong input or output. 
Also the whole condition is stringified and will be longer than these two as separate asserts by adding ` && ` 😄.